Objects/codeobject.c

```diff
@@ -701,13 +703,68 @@ PyCode_NewEmpty(const char *filename, const char *funcname, int firstlineno)
    lnotab_notes.txt for the details of the lnotab representation.
 */
 
+
+int
+_PyCode_CreateLineArray(PyCodeObject *co)
+{
+    assert(co->_co_linearray == NULL);
+    PyCodeAddressRange bounds;
+    int size;
+    int max_line = 0;
+    _PyCode_InitAddressRange(co, &bounds);
+    while (1) {
+        if (!_PyLineTable_NextAddressRange(&bounds)) {
+            break;
+        }
+        if (bounds.ar_line > max_line) {
+            max_line = bounds.ar_line;
+        }
+    }
+    if (max_line < (1 << 15)) {
+        size = 2;
+    }
+    else {
+        size = 4;
+    }
+    co->_co_linearray = PyMem_Malloc(Py_SIZE(co)*size);
+    if (co->_co_linearray == NULL) {
+        PyErr_NoMemory();
+        return -1;
+    }
+    co->_co_linearray_entry_size = size;
+    _PyCode_InitAddressRange(co, &bounds);
+    while (1) {
+        if (!_PyLineTable_NextAddressRange(&bounds)) {
+            break;
+        }
+        int addr = bounds.ar_start;
+        while (addr < bounds.ar_end) {
+            assert(addr < (int)(Py_SIZE(co) * sizeof(_Py_CODEUNIT)));
+            int index = addr / sizeof(_Py_CODEUNIT);
+            if (size == 2) {
+                assert(((int16_t)bounds.ar_line) == bounds.ar_line);
+                ((int16_t *)co->_co_linearray)[index] = bounds.ar_line;
+            }
+            else {
+                assert(size == 4);
+                ((int32_t *)co->_co_linearray)[index] = bounds.ar_line;
+            }
+            addr += sizeof(_Py_CODEUNIT);
+        }
```

Comment:
Minor: I find this easier to follow as a for-loop:

```suggestion
        int j = bounds.ar_end / sizeof(_Py_CODEUNIT);
        for (int i = bounds.ar_start / sizeof(_Py_CODEUNIT); i < j; i++) {
            assert(i < Py_SIZE(co));
            if (size == 2) {
                assert(((int16_t)bounds.ar_line) == bounds.ar_line);
                ((int16_t *)co->_co_linearray)[i] = bounds.ar_line;
            }
            else {
                assert(size == 4);
                ((int32_t *)co->_co_linearray)[i] = bounds.ar_line;
            }
        }
```